src/Data/Parameterized/MonoidSolver.hs

```diff
@@ -0,0 +1,283 @@
+-- expressions, corresponds to the bracket operator in the paper.
+type family Diff (me :: MonoidExpr m) :: MonoidExpr m ~> MonoidExpr m where
+  Diff (me1 :<>: me2) = ComposeDiffSym2 me1 me2
+  Diff 'EUnit = IdSym1
+  Diff ('EVar m) = EOpSym1 ('EVar m)
+
+-- | 'UnDiff' corresponds to "reify" from the paper.
+type family UnDiff (diff :: MonoidExpr m ~> MonoidExpr m) :: MonoidExpr m where
+  UnDiff diff = Apply diff 'EUnit
+
+type family Normalize (me :: MonoidExpr m) :: MonoidExpr m where
+  Normalize me = UnDiff (Diff me)
+
+-- | Note on termination: The one recursive call to this function is strictly
+-- decreasing on its first argument. This function is mutually recursive with
+-- 'normalizeSound' and recursive calls to that function are made on the second
+-- argument, but calls to this function *from* 'normalizeSound' are on strict
+-- subexpressions of its argument.
+normalizeLemma ::
+  forall k me1 me2.
+  TypeLevelMonoid k =>
+  MERepr k me1 ->
+  MERepr k me2 ->
+  Eval k (Apply (Diff me1) (Apply (Diff me2) 'EUnit)) :~:
+    Op k (Eval k me1) (Eval k me2)
+normalizeLemma mer1 mer2 =
+  case mer1 of
+    EUnitRepr ->
+      case idl (Proxy :: Proxy (Eval k me2)) of
+        Refl ->
+          case norm mer2 of
+            Refl -> Refl
+    EVarRepr {} ->
+      case norm mer2 of
+        Refl -> Refl
+    EOpRepr (mer1' :: MERepr k me1') (mer2' :: MERepr k me2') ->
+      case normalizeLemma mer1' (EOpRepr mer2' mer2) of
+        Refl ->
+              assoc (Proxy :: Proxy (Eval k me1')) (Proxy :: Proxy (Eval k me2')) (Proxy :: Proxy (Eval k me2))
+  where
+    norm :: MERepr k me -> Eval k (Normalize me) :~: Eval k me
+    norm = normalizeSound
+
+normalizeSound ::
+  TypeLevelMonoid k =>
+  MERepr k me ->
+  Eval k (Normalize me) :~: Eval k me
+normalizeSound =
+  \case
+    EUnitRepr -> Refl
+    EVarRepr (sing :: Proxy n) -> idr sing
+    EOpRepr (mer1 :: MERepr k me1) (mer2 :: MERepr k me2) ->
+      normalizeLemma mer1 mer2
+
+-- | Because of the construction of 'Normalize', the equality constraint always
+-- holds between 'MERepr' terms that are convertible with the monoid laws.
+solve ::
+  ( TypeLevelMonoid k
+  , Eval k (Normalize me1) ~ Eval k (Normalize me2)
+  )
+  =>
+  MERepr k me1 ->
+  MERepr k me2 ->
+  Eval k me1 :~: Eval k me2
+solve repr1 repr2 =
+  case (normalizeSound repr1, normalizeSound repr2) of
+    (Refl, Refl) -> Refl
+
+----------------------------------------------------------------------
+-- KnownRepr
+--
+
+instance KnownRepr (MERepr k) 'EUnit where
+  knownRepr = EUnitRepr
+
+instance KnownRepr (MERepr k) ('EVar n) where
+  knownRepr = EVarRepr (Proxy :: Proxy n)
+
+instance
+  ( KnownRepr (MERepr k) me1
+  , KnownRepr (MERepr k) me2
+  ) => KnownRepr (MERepr k) (me1 ':<>: me2) where
+  knownRepr = EOpRepr knownRepr knownRepr
+
+----------------------------------------------------------------------
+-- Instances
+--
+
+type instance Op Nat n1 n2 = n1 + n2
```

Comment:
Now that you have a `TypeLevelMonoid` class, would it make more sense to have `Op` and `Unit` be associated type families in `TypeLevelMonoid`? Unless I'm missing something, any time you'd define instances for `Op` and `Unit`, you'd also want to define a corresponding instance for `TypeLevelMonoid` as well, so better to make this relationship explicit.

Reply:
Yes, good point!